src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcClient.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipes;
+using System.Linq;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using System.Security.Principal;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    internal class IpcClient
+    {
+        private static string IpcRootPath { get; } = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? @"\\.\pipe\" : Path.GetTempPath();
+
+        private static double ConnectTimeoutMilliseconds { get; } = TimeSpan.FromSeconds(3).TotalMilliseconds;
+
+        /// <summary>
+        /// Get the OS Transport to be used for communicating with a dotnet process.
+        /// </summary>
+        /// <param name="processId">The PID of the dotnet process to get the transport for</param>
+        /// <returns>A System.IO.Stream wrapper around the transport</returns>
+        private static Stream GetTransport(int processId)
+        {
+            try 
+            {
+                var process = Process.GetProcessById(processId);
+            }
+            catch (System.ArgumentException)
+            {
+                throw new ServerNotAvailableException($"Process {processId} is not running.");
+            }
+ 
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                string pipeName = $"dotnet-diagnostic-{processId}";
+                var namedPipe = new NamedPipeClientStream(
+                    ".", pipeName, PipeDirection.InOut, PipeOptions.None, TokenImpersonationLevel.Impersonation);
+                namedPipe.Connect((int)ConnectTimeoutMilliseconds);
+                return namedPipe;
+            }
+            else
+            {
+                string ipcPort;
+                try
+                {
+                    ipcPort = Directory.GetFiles(IpcRootPath, $"dotnet-diagnostic-{processId}-*-socket") // Try best match.
+                                .OrderByDescending(f => new FileInfo(f).LastWriteTime)
+                                .FirstOrDefault();
+                    if (ipcPort == null)
+                    {
+                        throw new ServerNotAvailableException($"Process {processId} not running compatible .NET Core runtime.");
+                    }
+                }
+                catch (InvalidOperationException)
+                {
+                    throw new ServerNotAvailableException($"Process {processId} not running compatible .NET Core runtime.");
+                }
+                string path = Path.Combine(IpcRootPath, ipcPort);
+                var remoteEP = new UnixDomainSocketEndPoint(path);
```

Comment:
I'm pretty certain this API is `netstandard2.1`, so we'll have to do something about this if we want to target older runtimes or netstandards.  I don't think this is in the scope of this PR though 😄